scrapyd_k8s/log_handler.py

```diff
@@ -0,0 +1,161 @@
+import os
+import threading
+import tempfile
+import logging
+
+from kubernetes import client, watch
+from .object_storage import LibcloudObjectStorage
+
+logger = logging.getLogger(__name__)
+
+watcher_threads = {}
+pod_tmp_mapping = {}
+
+def get_last_n_lines(file_path, num_lines):
+    """Efficiently retrieve the last `num_lines` lines from `file_path`."""
+    try:
+        with open(file_path, 'rb') as f:
+            # Move to the end of the file
+            f.seek(0, os.SEEK_END)
+            file_size = f.tell()
+            block_size = 6144
+            data = b''
+            remaining_lines = num_lines
+            while remaining_lines > 0 and file_size > 0:
+                if file_size - block_size > 0:
+                    f.seek(file_size - block_size)
+                    block_data = f.read(block_size)
+                else:
+                    f.seek(0)
+                    block_data = f.read(file_size)
+                file_size -= block_size
+                data = block_data + data
+                lines_found = data.count(b'\n')
+                # If we've found enough lines, we can stop reading more blocks
+                if lines_found >= remaining_lines:
+                    break
+
+            # Decode the data and split into lines
+            lines = data.decode('utf-8', errors='replace').splitlines()
+            # Return the last `num_lines`
+            result = lines[-num_lines:]
+            return result
+    except FileNotFoundError:
+        logger.warning(f"File not found: {file_path}")
+        return []
+
+def concatenate_and_delete_files(main_file_path, temp_file_path):
+    try:
+        with open(main_file_path, 'a') as main_file, open(temp_file_path, 'r') as temp_file:
+            main_file.write(temp_file.read())
```

Comment:
Would this read the whole file into memory?
If so, please do it by chunks to avoid a memory hog (and greatly increasing the required memory resources to run scrapyd-k8s).

Reply:
Yes, thank you for catching this.